### ISSUE TITLE
BrowserComponent doesn't show PDF in iOS

### ISSUE THREAD

**Issue by schreiby** [codenameone/CodenameOne#2597]
Hi, 

I understand that PDFs won't display properly on the iOS **simulator** when using BrowserComponent  (they come up as blank white spaces),  but it's not working on a **real iPhone** either -- any PDF that I want to display also comes up as a blank, white space, just like on the simulator. 

Here's my code: 

`File hello = new File(FileSystemStorage.getInstance().getAppHomePath()+ fileToUpload.getFileName());`
`Log.p("hello: " + hello.getAbsolutePath());`
 `// fileToUpload.getLocation() is a valid Storage location`
`InputStream is = Storage.getInstance().createInputStream(fileToUpload.getLocation());`
`OutputStream os = FileSystemStorage.getInstance().openOutputStream(hello.getAbsolutePath());`

`Util.copy(is, os);`
`is.close();`
`os.close();`
`Util.cleanup(is);`
`Util.cleanup(os);`

`BrowserComponent browserComponent = new BrowserComponent();`
`browserComponent.setURL(hello.getAbsolutePath());`
`horizontalContainer.add(browserComponent);`

Thanks!

**Comment by shannah** [comment]
That won't work.  You can't pass file paths to setURL.  You'll have better luck with Display.getInstance().execute(hello.getAbsolutePath());

Or you can try using the webserver cn1lib (https://github.com/shannah/CN1Webserver)

A better place for this type of inquiry is either the mailing list or Stack overflow

**Comment by shannah** [comment]
My mistake.  Actually file URLs do work in iOS.  Here is a test case that I created that correctly displays a PDF in the webview on iOS.

~~~~
private void testViewPDFInWebview() {
        Form hi = new Form("PDF View", new BorderLayout());
        BrowserComponent bc = new BrowserComponent();
        File file = new File("scansmpl.pdf");
        FileSystemStorage fs = FileSystemStorage.getInstance();
        
        try {
            OutputStream os = fs.openOutputStream(file.getAbsolutePath());
            InputStream is = Display.getInstance().getResourceAsStream(null, "/scansmpl.pdf");
            
            Util.copy(is, os);
            
            bc.setURL(file.getAbsolutePath());
            hi.add(BorderLayout.CENTER, bc);
            hi.show();
            
        } catch (Throwable t) {
            Log.e(t);
        }
    }
~~~~

A screenshot of it running in iOS 

![image](https://user-images.githubusercontent.com/2677562/47742134-37126880-dc39-11e8-8c08-2d3071fdd815.png)

The sample PDF file I used.  Place this in the "src" directory of your project.  The test copies this into file system storage and then loads the copy in the webview to demonstrate file URLs working.

[scansmpl.pdf](https://github.com/codenameone/CodenameOne/files/2531135/scansmpl.pdf)

If your sample isn't working, focus on the path that you are passing into getURL() and make sure that the file exists.

**Comment by shannah** [comment]
One other thing that occurs to me now, looking at your code.

~~~~
InputStream is = Storage.getInstance().createInputStream(fileToUpload.getLocation());
OutputStream os = FileSystemStorage.getInstance().openOutputStream(hello.getAbsolutePath());

Util.copy(is, os);
~~~~

Some platforms just store Storage in the home directory (I think iOS does this also).   So you may be copying a file to itself in this case, which might be obliterating it.  Try making the destination file name different than the source file name.

**Comment by schreiby** [comment]
Thankfully this issue has been solved-- if the file name has a space, the BrowserComponent wouldn't show it. I also added this information to the [SO question](https://stackoverflow.com/a/53103379/5095571) so it'll be accessible to anyone who runs into this issue.

Thanks so much for your help!

**Comment by shannah** [comment]
I have just added a `toURL()` method in the `File` class so that you'll be able to more reliably open files from the file system in a browser component.

Use `bc.setURL(myFile.toURL().toString())` instead of `bc.setURL(myFile.getAbsolutePath())` to ensure that it is a valid URL (this will encode spaces in file names correctly).

I also added convenience methods to just pass a URL to setURL instead of a string.  E.g. `bc.setURL(myFile.toURL())`
